mmcls/models/retriever/image2image.py

```diff
@@ -0,0 +1,298 @@
+        """Compare the prototype and calculate the similarity.
+
+        Args:
+            inputs (torch.Tensor): The input tensor with shape (N, C).
+            fn (Callable): Function to calculate the similarity between
+                the extracted feature of inputs and the prototype.
+            data_samples (List[BaseDataElement], optional): The annotation
+                data of every samples. Defaults to None.
+        Returns:
+            dict: a dictionary of score and prediction label based on fn.
+        """
+        sim, indices = fn(inputs, self.prototype_vecs)
+        predictions = {'score': sim, 'pred_label': indices}
+        return predictions
+
+    def predict(self,
+                inputs: tuple,
+                data_samples: Optional[List[ClsDataSample]] = None,
+                **kwargs) -> List[ClsDataSample]:
+        """Predict results from the extracted features.
+
+        Args:
+            inputs (tuple): The features extracted from the backbone.
+            data_samples (List[ClsDataSample], optional): The annotation
+                data of every samples. Defaults to None.
+            **kwargs: Other keyword arguments accepted by the ``predict``
+                method of :attr:`head`.
+        Returns:
+            List[ClsDataSample]: the raw data_samples with
+                the predicted results
+        """
+        if not self.prototype_inited:
+            self.prepare_prototype()
+
+        feats = self.extract_feat(inputs)
+        if isinstance(feats, tuple):
+            feats = feats[-1]
+
+        # Matching of similarity
+        result = self.matching(
+            feats, lambda query, gallery: torch.sort(
+                torch.cosine_similarity(
+                    query.unsqueeze(1), gallery.unsqueeze(0), dim=-1),
+                dim=-1,
+                descending=True))
+
+        pred_scores = result['score']
+        pred_labels = result['pred_label']
+        if data_samples is not None:
+            for data_sample, score, label in zip(data_samples, pred_scores,
+                                                 pred_labels):
+                data_sample.set_pred_score(score).set_pred_label(label)
+        else:
+            data_samples = []
+            for score, label in zip(pred_scores, pred_labels):
+                data_samples.append(ClsDataSample().set_pred_score(
+                    score).set_pred_label(label))
+        self.post_process(data_samples)
+        return data_samples
+
+    def __get_prototype_from_dataloader(self, device):
+        data_loader = self.prototype
+        num = len(data_loader.dataset)
+
+        self.prototype_vecs = torch.zeros(num, self.dim)
+        self.prototype_vecs = self.prototype_vecs.to(device)
+
+        for data in data_loader:
+            batch_num = len(data['inputs'])
+            data = self.data_preprocessor(data, False)
+
+            if isinstance(data, dict):
```

Comment:
data output from mmcls.data_preprocessor should be a dict, refert to https://github.com/open-mmlab/mmclassification/blob/8c5d86a388e71d429851da702c4b92ccb9b90f87/mmcls/models/utils/data_preprocessor.py#L79. It doesn't need to be so complicated here.